### PR TITLE
Various fixes for bokeh static_source optimization

### PR DESCRIPTION
**Pull request by philippjfr**
In https://github.com/ioam/holoviews/issues/1396 it was suggested that the ``static_source`` optimization in bokeh should only kick in if both the element and its data are static.

### CODE REVIEW

**Comment by jlstevens** [comment]
Looks good.

I can't really see how to test this right now so I'll merge once the tests pass.

**Comment by philippjfr** [comment]
So this approach does not work since it just disables the optimization entirely. The issue is twofold a) when a custom style is applied the element id changes, b) something in the plotting pipeline applies a dynamic method to the data, which again ends up changing the element id. Therefore the plotting code never sees the same element id twice and the optimization does nothing. Will have to think about this.

**Comment by jlstevens** [comment]
Problem b) seems like the real issue to me as I don't think disabling the optimization for a) means it will always be disabled.

File: holoviews/core/dimension.py [hunk 1]
```diff
@@ -488,6 +489,7 @@ def __init__(self, data, id=None, **params):
         """
         self.data = data
         self.id = id
+        self._plot_id = plot_id or obj_id(self)
```
Comment (jlstevens) [inline holoviews/core/dimension.py]: Just use ``__builtins__.id`` to avoid the name shadowing issue.

File: holoviews/core/dimension.py [hunk 2]
```diff
@@ -488,6 +489,7 @@ def __init__(self, data, id=None, **params):
         """
         self.data = data
         self.id = id
```
Comment (jlstevens) [inline holoviews/core/dimension.py]: Do we ever actually set the ``id`` via the constructor?

Comment (jlstevens) [inline holoviews/core/dimension.py]: Ah, probably for clone. Nevermind my question!

File: holoviews/plotting/bokeh/element.py [hunk 3]
```diff
@@ -804,15 +804,16 @@ def update_frame(self, key, ranges=None, plot=None, element=None, empty=False):
         # Cache frame object id to skip updating data if unchanged
         previous_id = self.handles.get('previous_id', None)
         if self.batched:
-            current_id = sum(element.traverse(lambda x: id(x.data), [Element]))
+            current_id = sum(element.traverse(lambda x: x._plot_id, [Element]))
         else:
-            current_id = id(element.data)
+            current_id = element._plot_id
```
Comment (jlstevens) [inline holoviews/plotting/bokeh/element.py]: Is it no longer taking into account the ``id`` of the data? It seems to do that in tabular...

File: holoviews/plotting/bokeh/plot.py [hunk 4]
```diff
@@ -147,7 +147,7 @@ def sync_sources(self):
         from the same object.
         """
         get_sources = lambda x: (id(x.current_frame.data), x)
-        filter_fn = lambda x: (x.shared_datasource and x.current_frame and
+        filter_fn = lambda x: (x.shared_datasource and x.current_frame is not None and
```
Comment (jlstevens) [inline holoviews/plotting/bokeh/plot.py]: Just to confirm: is this a a fix unrelated to the optimization itself?

File: holoviews/plotting/bokeh/tabular.py [hunk 5]
```diff
@@ -73,11 +73,12 @@ def current_handles(self):
         if self.static and not self.dynamic:
             return handles
 
+
+        element = self.current_frame
         previous_id = self.handles.get('previous_id', None)
-        current_id = id(self.current_frame.data) if self.current_frame else None
+        current_id = None if self.current_frame is None else id(element)+id(element.data)
```
Comment (jlstevens) [inline holoviews/plotting/bokeh/tabular.py]: Summing ``ids`` is not a safe test for uniqueness as many pairs of ids could sum to the same value (even if it is unlikely!). Note that collisions might not even be *that* rare as the ``id`` doesn't return a hash - it is more like  to a memory address.

To check the identity of both the element and the data I would use the tuple ``(id(element), id(element.data))`` and not the sum.

File: holoviews/core/dimension.py (hunk 2 above)
Comment (philippjfr) [inline holoviews/core/dimension.py]: Yes, clone does all the time.

File: holoviews/plotting/bokeh/plot.py (hunk 4 above)
Comment (philippjfr) [inline holoviews/plotting/bokeh/plot.py]: Seems so, not sure why I pushed it here, but it is an important fix.

**Comment by jlstevens** [comment]
Looks good with the latest fixes. One last import issue to fix and then I'll merge once the tests go green. That said, are there any meaningful unit tests we should set up to test the optimization?

**Comment by jlstevens** [comment]
Thanks for adding the unit tests! Merging.

**Pull request merged**